graph/api/src/main/scala/org/apache/spark/graph/api/CypherSession.scala

```diff
@@ -0,0 +1,258 @@
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * Note that queries can take optional parameters:
+   *
+   * {{{
+   *     Parameters:
+   *
+   *     {
+   *        "name" : "Alice"
+   *     }
+   *
+   *     Query:
+   *
+   *     MATCH (n:Person)
+   *     WHERE n.name = $name
+   *     RETURN n
+   * }}}
+   *
+   * @param graph      [[PropertyGraph]] on which the query is executed
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(graph: PropertyGraph, query: String, parameters: Map[String, Any]): CypherResult
+
+  /**
+   * Executes a Cypher query on the given input graph.
+   *
+   * Note that queries can take optional parameters:
+   *
+   * {{{
+   *     Parameters:
+   *
+   *     {
+   *        "name" : "Alice"
+   *     }
+   *
+   *     Query:
+   *
+   *     MATCH (n:Person)
+   *     WHERE n.name = $name
+   *     RETURN n
+   * }}}
+   *
+   * @param graph      [[PropertyGraph]] on which the query is executed
+   * @param query      Cypher query to execute
+   * @param parameters parameters used by the Cypher query
+   * @since 3.0.0
+   */
+  def cypher(
+      graph: PropertyGraph,
+      query: String,
+      parameters: java.util.Map[String, Object]): CypherResult = {
+    cypher(graph, query, parameters.asScala.toMap)
+  }
+
+  /**
+   * Creates a [[PropertyGraph]] from a sequence of [[NodeFrame]]s and [[RelationshipFrame]]s.
+   * At least one [[NodeFrame]] has to be provided.
+   *
+   * For each label set and relationship type there can be at most one [[NodeFrame]] and at most one
+   * [[RelationshipFrame]], respectively.
+   *
+   * @param nodes         NodeFrames that define the nodes in the graph
+   * @param relationships RelationshipFrames that define the relationships in the graph
+   * @since 3.0.0
+   */
+  def createGraph(nodes: Array[NodeFrame], relationships: Array[RelationshipFrame]): PropertyGraph
+
+  /**
+   * Creates a [[PropertyGraph]] from nodes and relationships.
+   *
+   * The given dataset needs to adhere to the following column naming conventions:
+   *
+   * {{{
+   *     Id column:        `$ID`            (nodes and relationships)
+   *     SourceId column:  `$SOURCE_ID`     (relationships)
+   *     TargetId column:  `$TARGET_ID`     (relationships)
+   *
+   *     Label columns:    `:{LABEL_NAME}`  (nodes)
+   *     RelType columns:  `:{REL_TYPE}`    (relationships)
+   *
+   *     Property columns: `{Property_Key}` (nodes and relationships)
+   * }}}
+   *
+   * @note It is recommended to cache the input datasets if they represent multiple label sets and
+   *       relationship types.
+   *
+   * @see [[CypherSession]]
+   * @param nodes         node dataset
+   * @param relationships relationship dataset
+   * @since 3.0.0
+   */
+  def createGraph(nodes: Dataset[Row], relationships: Dataset[Row]): PropertyGraph = {
```

Comment:
Let's move the function logic to `object CyperSession` by making a helper function.

Reply:
How do you propose we inject the `Logging` instance to still be able to log the warning?

Reply:
That is not the main reason to ask this. The logic is really helper logic. We had better move that out to this `trait`.

Reply:
I agree.